src/shims/posix/foreign_items.rs

```diff
@@ -55,8 +55,13 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
             // File related shims
             "open" | "open64" => {
-                let &[ref path, ref flag, ref mode] = this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
-                let result = this.open(path, flag, mode)?;
+                // `open` is variadic, the third argument is only present when the second argument has O_CREAT (or on linux O_TMPFILE, but miri doesn't support that) set
+                this.check_abi_and_shim_symbol_clash(abi, Abi::C { unwind: false }, link_name)?;
+                if args.len() < 2 || args.len() > 3 {
+                    throw_ub_format!("incorrect number of arguments: got {}, expected 2 or 3", args.len());
+                }
```

Comment:
I think I would prefer this check to happen inside `open`, similar to how we handle `fcntl`. This file here should have as little function-specific logic as possible.